src/Engines/ElasticsearchEngine.php

```diff
@@ -0,0 +1,189 @@
+<?php
+
+namespace Laravel\Scout\Engines;
+
+use Laravel\Scout\Builder;
+use Elasticsearch\Client as Elasticsearch;
+use Illuminate\Database\Eloquent\Collection;
+use Illuminate\Support\Collection as BaseCollection;
+
+class ElasticsearchEngine extends Engine
+{
+    /**
+     * @var string $index
+     */
+    protected $index;
+
+    /**
+     * Create a new engine instance.
+     *
+     * @param  \Elasticsearch\Client  $elasticsearch
+     * @return void
+     */
+    public function __construct(Elasticsearch $elasticsearch, $index)
+    {
+        $this->elasticsearch = $elasticsearch;
+
+        $this->index = $index;
+    }
+
+    /**
+     * Update the given model in the index.
+     *
+     * @param  Collection  $models
+     * @return void
+     */
+    public function update($models)
+    {
+        $models->each(function ($model) {
+            $this->elasticsearch->index([
+                'index' => $this->index,
+                'type' => $model->searchableAs(),
+                'id' => $model->getKey(),
+                'body' => $model->toSearchableArray(),
+            ]);
+        });
+    }
+
+    /**
+     * Remove the given model from the index.
+     *
+     * @param  Collection  $models
+     * @return void
+     */
+    public function delete($models)
+    {
+        $models->each(function ($model) {
+            $this->elasticsearch->delete([
+                'index' => $this->index,
+                'type' => $model->searchableAs(),
+                'id'  => $model->getKey(),
+            ]);
+        });
+    }
+
+    /**
+     * Perform the given search on the engine.
+     *
+     * @param  Builder  $query
+     * @return mixed
+     */
+    public function search(Builder $query)
+    {
+        return $this->performSearch($query, array_filter([
```

Comment:
Correct me if I'm wrong, but I find this array_filter() pointless because we know the values are not null. performSearch already takes care of this, if this stays some performSearch conditions are pointless